File: forge-harness/src/main/java/forge/harness/common/ChoiceSpace.java
package forge.harness.common;

import forge.game.card.CardCollection;
import forge.game.card.CardCollectionView;
import forge.util.collect.FCollectionView;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public final class ChoiceSpace {

    public static final class ChoiceResult<T> {
        public final List<T> sorted;
        public final int index;
        public final T chosen;

        private ChoiceResult(final List<T> sorted, final int index, final T chosen) {
            this.sorted = sorted;
            this.index = index;
            this.chosen = chosen;
        }

        public boolean isPass() {
            return index >= sorted.size();
        }
    }

    private ChoiceSpace() {}

    public static <T> T pickOne(final FCollectionView<T> options, final Random rng) {
        if (options == null || options.isEmpty()) {
            return null;
        }
        if (options.size() == 1) {
            ParityLog.log("pick_one", 1, "idx=0");
            return options.get(0);
        }
        final int idx = rng.nextInt(options.size());
        ParityLog.log("pick_one", options.size(), "idx=" + idx);
        return options.get(idx);
    }

    public static <T> T pickOne(final List<T> options, final Random rng) {
        if (options == null || options.isEmpty()) {
            return null;
        }
        if (options.size() == 1) {
            ParityLog.log("pick_one", 1, "idx=0");
            return options.get(0);
        }
        final int idx = rng.nextInt(options.size());
        ParityLog.log("pick_one", options.size(), "idx=" + idx);
        return options.get(idx);
    }

    public static int pickCount(final int min, final int max, final int available, final Random rng) {
        final int hi = Math.min(Math.max(max, 0), Math.max(available, 0));
        final int lo = Math.min(Math.max(min, 0), hi);
        final int count = lo + (hi > lo ? rng.nextInt(hi - lo + 1) : 0);
        ParityLog.log("pick_count [" + min + "-" + max + "]", available, String.valueOf(count));
        return count;
    }

    public static int pickIntInRange(final int min, final int max, final Random rng) {
        if (max <= min) {
            ParityLog.log("pick_int_in_range[" + min + " to " + max + "]", null, String.valueOf(min));
            return min;
        }
        final long span = (long) max - (long) min + 1L;
        final int val;
        if (span <= Integer.MAX_VALUE) {
            val = min + rng.nextInt((int) span);
        } else {
            final long candidate = (long) min + rng.nextInt(Integer.MAX_VALUE);
            val = (int) Math.min(candidate, (long) max);
        }
        ParityLog.log("pick_int_in_range [" + min + " to " + max + "]", null, String.valueOf(val));
        return val;
    }

    public static CardCollection pickManyCards(
            final CardCollectionView options,
            final int min,
            final int max,
            final Random rng) {
        final CardCollection pool = new CardCollection(options == null ? new CardCollection() : options);
        final int count = pickCount(min, max, pool.size(), rng);
        final CardCollection out = new CardCollection();
        for (int i = 0; i < count && !pool.isEmpty(); i++) {
            out.add(pool.remove(pickIndex(pool.size(), rng)));
        }
        final int poolSize = options == null ? 0 : options.size();
        ParityLog.log("pick_many_unique [" + min + " to " + max + "]", poolSize, "picked " + out.size());
        return out;
    }

    public static <T> List<T> shuffleCopy(final List<T> options, final Random rng) {
        final List<T> out = new ArrayList<>(options);
        Collections.shuffle(out, rng);
        ParityLog.log("shuffle_copy", options.size(), "done");
        return out;
    }

    public static boolean pickBool(final Random rng) {
        final boolean result = rng.nextInt(2) == 1;
        ParityLog.log("pick_bool", 2, String.valueOf(result));
        return result;
    }

    public static int pickIndex(final int size, final Random rng) {
        if (size <= 0) {
            ParityLog.log("pick_index", 0, "idx=-1");
            return -1;
        }
        if (size == 1) {
            ParityLog.log("pick_index", 1, "idx=0");
            return 0;
        }
        final int idx = rng.nextInt(size);
        ParityLog.log("pick_index", size, "idx=" + idx);
        return idx;
    }

    public static int pickIndexWithPass(final int size, final Random rng) {
        if (size < 0) {
            return 0;
        }
        final int idx = rng.nextInt(size + 1);
        final String outcome = idx >= size ? "PASS" : "idx=" + idx;
        ParityLog.log("pick_index_with_pass", size, outcome);
        return idx;
    }

    public static <T> List<T> sortNative(
            final List<T> nativeOptions,
            final Comparator<T> parityComparator
    ) {
        final List<T> sorted = new ArrayList<>(nativeOptions);
        sorted.sort(parityComparator);
        return sorted;
    }

    public static <T> ChoiceResult<T> chooseFromNative(
            final List<T> nativeOptions,
            final Comparator<T> parityComparator,
            final boolean allowPass,
            final Random rng
    ) {
        final List<T> sorted = sortNative(nativeOptions, parityComparator);
        if (sorted.isEmpty()) {
            return new ChoiceResult<>(sorted, allowPass ? 0 : -1, null);
        }
        final int idx = allowPass ? pickIndexWithPass(sorted.size(), rng) : pickIndex(sorted.size(), rng);
        final T chosen = idx >= sorted.size() ? null : sorted.get(idx);
        return new ChoiceResult<>(sorted, idx, chosen);
    }

    public static int pickWeightedIndexWithPass(final int size, final int actionWeight, final Random rng) {
        final int safeSize = Math.max(0, size);
        final int safeWeight = Math.max(1, actionWeight);
        final int totalWeight = safeSize * safeWeight + 1;
        final int roll = rng.nextInt(totalWeight);
        final int idx;
        if (roll >= safeSize * safeWeight) {
            idx = safeSize;
        } else {
            idx = roll / safeWeight;
        }
        final String outcome = idx >= safeSize ? "PASS" : "idx=" + idx;
        ParityLog.log("pick_weighted w=" + safeWeight, safeSize, outcome);
        return idx;
    }
}

File: manabrew-rs/crates/parity/src/choice_space.rs
use crate::java_random::JavaRandom;
use crate::parity_log;
use crate::protocol::ChoiceLogEntry;
use std::cmp::Ordering;

fn rng_log(name: &str, choices: Option<usize>, outcome: String, rng: &JavaRandom) {
    parity_log::log(ChoiceLogEntry {
        name: name.to_string(),
        choices,
        outcome,
        rng_call_count: Some(rng.call_count),
    });
}

// ── Choice functions ──────────────────────────────────────────────

pub fn pick_bool(rng: &mut JavaRandom) -> bool {
    let idx = pick_index_inner(2, rng);
    let result = idx == 1;
    rng_log("pick_bool", Some(2), result.to_string(), rng);
    result
}

pub fn pick_one<T: Copy>(options: &[T], rng: &mut JavaRandom) -> Option<T> {
    if options.is_empty() {
        return None;
    }
    let idx = pick_index_inner(options.len(), rng);
    rng_log("pick_one", Some(options.len()), format!("idx={idx}"), rng);
    Some(options[idx])
}

/// Pick index in [0, size). Does not consume RNG when size <= 1.
pub fn pick_index(size: usize, rng: &mut JavaRandom) -> usize {
    let idx = pick_index_inner(size, rng);
    rng_log("pick_index", Some(size), format!("idx={idx}"), rng);
    idx
}

/// Core index draw without logging. Used internally by wrappers (`pick_bool`,
/// `pick_one`, ...) so they emit only their own callback line — Java's
/// `ChoiceSpace.pickOne` likewise logs a single entry per call.
fn pick_index_inner(size: usize, rng: &mut JavaRandom) -> usize {
    if size <= 1 {
        return 0;
    }
    rng.next_int(size as i32) as usize
}

/// Pick index in [0, size] where size is PASS.
pub fn pick_index_with_pass(size: usize, rng: &mut JavaRandom) -> usize {
    let idx = rng.next_int((size + 1) as i32) as usize;
    let outcome = if idx >= size {
        "PASS".into()
    } else {
        format!("idx={idx}")
    };
    rng_log("pick_index_with_pass", Some(size), outcome, rng);
    idx
}

pub fn pick_weighted_index_with_pass(
    size: usize,
    action_weight: usize,
    rng: &mut JavaRandom,
) -> usize {
    let safe_size = size;
    let safe_weight = action_weight.max(1);
    let total = safe_size.saturating_mul(safe_weight).saturating_add(1);
    let roll = pick_index(total, rng);
    let idx = if roll >= safe_size.saturating_mul(safe_weight) {
        safe_size
    } else {
        roll / safe_weight
    };
    let outcome = if idx >= safe_size {
        "PASS".into()
    } else {
        format!("idx={idx}")
    };
    rng_log(
        &format!("pick_weighted w={safe_weight}"),
        Some(safe_size),
        outcome,
        rng,
    );
    idx
}

pub fn sort_native<T: Clone>(native: &[T], mut cmp: impl FnMut(&T, &T) -> Ordering) -> Vec<T> {
    let mut out = native.to_vec();
    out.sort_by(|a, b| cmp(a, b));
    out
}

pub fn pick_count(min: usize, max: usize, available: usize, rng: &mut JavaRandom) -> usize {
    let hi = max.min(available);
    let lo = min.min(hi);
    let count = lo
        + if hi > lo {
            rng.next_int((hi - lo + 1) as i32) as usize
        } else {
            0
        };
    rng_log(
        &format!("pick_count [{min}-{max}]"),
        Some(available),
        count.to_string(),
        rng,
    );
    count
}

pub fn pick_many_unique<T: Copy>(
    options: &[T],
    min: usize,
    max: usize,
    rng: &mut JavaRandom,
) -> Vec<T> {
    let mut pool = options.to_vec();
    let count = pick_count(min, max, pool.len(), rng);
    let mut out = Vec::new();
    for _ in 0..count {
        if pool.is_empty() {
            break;
        }
        let idx = pick_index(pool.len(), rng);
        out.push(pool.remove(idx));
    }
    let len = options.len();
    let picked = out.len();
    rng_log(
        &format!("pick_many_unique [{min}-{max}]"),
        Some(len),
        format!("picked {picked}"),
        rng,
    );
    out
}
